### ISSUE TITLE
[Inductor] Significant numerical divergence (4.7% relative error) in Conv2d on CPU with torch.compile

### ISSUE THREAD

**Issue by zifan6699** [pytorch/pytorch#173921]
### 🐛 Describe the bug

Description
When using `torch.compile` on CPU, a single `nn.Conv2d` layer produces significant numerical divergence compared to the CUDA Eager implementation. In a specific case with large input values (around 1e13), the relative error reaches **4.7%**, which far exceeds the expected precision for float32.

Repro Data
I have isolated the issue to a single layer with specific weights and inputs.
- **Max Absolute Error**: 1.0486e+06
- **Max Relative Error**: 4.7062e-02 (Threshold: 1e-4)
- **Layer**: nn.Conv2d(3, 32, kernel_size=3, stride=2, padding=1, bias=False)

```python
import torch
import torch.nn as nn
import os

def run():
    x, sd = torch.load("input.pt"), torch.load("sd.pt")
    layer = nn.Conv2d(3, 32, 3, stride=2, padding=1, bias=False)
    layer.load_state_dict(sd)
    layer.eval()

    # Mode A: CUDA Eager
    with torch.no_grad():
        out_a = layer.cuda()(x.cuda())

    # Mode B: CPU Compile
    os.environ["INDUCTOR_DYNAMIC"] = "1"
    with torch.no_grad():
        out_b = torch.compile(layer.cpu())(x.cpu())

    # Comparison
    a, b = out_a.cpu().float(), out_b.cpu().float()
    diff = (a - b).abs()
    max_abs = diff.max().item()
    max_rel = (diff / a.abs().clamp(min=1e-12)).max().item()

    print(f"Max Absolute Error: {max_abs:.4e}")
    print(f"Max Relative Error: {max_rel:.4e}")
    if max_rel > 1e-4:
        print("Conclusion: Relative Error Exceeds 1e-4")
    else:
        print("Conclusion: Relative Error Within 1e-4")
    print("="*50)

if __name__ == "__main__":
    run()

```
```
Max Absolute Error: 1.0486e+06
Max Relative Error: 4.7062e-02
Conclusion: Relative Error Exceeds 1e-4
==================================================
```

[repro_conv1_divergence.zip](https://github.com/user-attachments/files/24963646/repro_conv1_divergence.zip)

### Versions

PyTorch version: 2.6.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.9.23 (main, Jun  5 2025, 13:40:20)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.8.0-90-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.6.20
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090

Nvidia driver version: 560.35.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        40 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               48
On-line CPU(s) list:                  0-47
Vendor ID:                            GenuineIntel
Model name:                           QEMU Virtual CPU version 2.5+
CPU family:                           15
Model:                                107
Thread(s) per core:                   1
Core(s) per socket:                   48
Socket(s):                            1
Stepping:                             1
BogoMIPS:                             4190.15
Flags:                                fpu de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx lm constant_tsc nopl xtopology cpuid tsc_known_freq pni ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c hypervisor lahf_lm abm cpuid_fault pti bmi1 avx2 bmi2 avx512f avx512dq avx512cd avx512bw avx512vl
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            1.5 MiB (48 instances)
L1i cache:                            1.5 MiB (48 instances)
L2 cache:                             192 MiB (48 instances)
L3 cache:                             16 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-47
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          KVM: Mitigation: VMX unsupported
Vulnerability L1tf:                   Mitigation; PTE Inversion
Vulnerability Mds:                    Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown:               Mitigation; PTI
Vulnerability Mmio stale data:        Unknown: No mitigations
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Vulnerable
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Retpoline
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected
Vulnerability Vmscape:                Not affected

Versions of relevant libraries:
[pip3] numpy==2.0.2
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] onnx==1.19.1
[pip3] onnxruntime==1.19.2
[pip3] open_clip_torch==3.2.0
[pip3] pytorch-lightning==0.7.1
[pip3] torch==2.6.0+cu126
[pip3] torch-geometric==2.6.1
[pip3] torchaudio==2.6.0+cu126
[pip3] torchvision==0.21.0+cu126
[pip3] triton==3.2.0
[conda] numpy                             2.0.2            pypi_0           pypi
[conda] nvidia-cublas-cu12                12.6.4.1         pypi_0           pypi
[conda] nvidia-cuda-cupti-cu12            12.6.80          pypi_0           pypi
[conda] nvidia-cuda-nvrtc-cu12            12.6.77          pypi_0           pypi
[conda] nvidia-cuda-runtime-cu12          12.6.77          pypi_0           pypi
[conda] nvidia-cudnn-cu12                 9.5.1.17         pypi_0           pypi
[conda] nvidia-cufft-cu12                 11.3.0.4         pypi_0           pypi
[conda] nvidia-curand-cu12                10.3.7.77        pypi_0           pypi
[conda] nvidia-cusolver-cu12              11.7.1.2         pypi_0           pypi
[conda] nvidia-cusparse-cu12              12.5.4.2         pypi_0           pypi
[conda] nvidia-cusparselt-cu12            0.6.3            pypi_0           pypi
[conda] nvidia-nccl-cu12                  2.21.5           pypi_0           pypi
[conda] nvidia-nvjitlink-cu12             12.6.85          pypi_0           pypi
[conda] nvidia-nvtx-cu12                  12.6.77          pypi_0           pypi
[conda] open-clip-torch                   3.2.0            pypi_0           pypi
[conda] pytorch-lightning                 0.7.1            pypi_0           pypi
[conda] torch                             2.6.0+cu126      pypi_0           pypi
[conda] torch-geometric                   2.6.1            pypi_0           pypi
[conda] torchaudio                        2.6.0+cu126      pypi_0           pypi
[conda] torchvision                       0.21.0+cu126     pypi_0           pypi
[conda] triton                            3.2.0            pypi_0           pypi

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by ayushsatyam146** [comment]
Hi @zifan6699 I agree on point number 1 and 2 and I think that's the best way to approach it. 
Regarding point number 3 I think you can achieve similar results via the below 
```python
# Mode A: CUDA Eager
    with torch.no_grad():
        # out_a = layer.cuda()(x.cuda())
        CL = torch.channels_last
        out_a = layer.cuda().to(memory_format=CL)(x.cuda().to(memory_format=CL))
```
I will really appreciate @Lucaskabela's review on this before I go ahead on implementing something like `torch.compile(mode='strict_numerical')`

**Comment by Cheng-Logic** [comment]
[PHYSICS_REJECT] The 4.7% relative divergence is a catastrophic cancellation artifact strictly working as intended; injecting artificially extreme inputs ($\sim 10^{11}$) into a highly parallelized convolution inherently exhausts the 23-bit mantissa of FP32, forcing severe absorption of smaller partial sums, meaning the memory layout shift (NCHW to NHWC) dictates disparate summation orderings that physically discard different segments of the data, a fundamental limitation of IEEE 754 arithmetic under extreme magnitude variance, not a compiler bug solvable by cuDNN updates.

**Comment by mlazos** [comment]
@ayushsatyam146 are you still working on this?

**Comment by hiahu329** [comment]
Update: I verified that this issue is still reproducible on both PyTorch 2.10.0+cu126 and PyTorch 2.12.1+cu126.

No repro files were changed. I used the same `173921.py`, `input.pt`, and `sd.pt`; the only change was switching the PyTorch conda environment/version.

Observed output is the same on both versions:

```text
Max Absolute Error: 1.0486e+06
Max Relative Error: 4.7062e-02
Conclusion: Relative Error Exceeds 1e-4
==================================================
